File: src/geometry/variableEvaluator.ts
import type {
  CadElement,
  ComputedGeometry,
  ComputedLine,
  ComputedVariable,
  DependencyError,
  ElementId,
  VariableElement
} from "../types/geometry";
import {
  dependencyError,
  geometryError,
  getPointAnchorOrError,
  numericError
} from "./evaluationContext";
import type { LocalVariableEvaluation } from "./evaluationContext";

const EPSILON = 1e-9;

const normalizeDegrees = (degrees: number) => (degrees + 360) % 360;

const computedLineOrError = (
  element: CadElement,
  lineId: ElementId,
  computedGeometry: Map<ElementId, ComputedGeometry>,
  elementsById: Map<ElementId, CadElement>,
  errors: DependencyError[],
  disabledByGroupId: Map<ElementId, ElementId>
): ComputedLine | undefined => {
  const line = computedGeometry.get(lineId);
  if (line?.kind !== "line") {
    errors.push(dependencyError(element, lineId, elementsById, disabledByGroupId));
    return undefined;
  }
  return line;
};

export const evaluateVariableElement = (
  element: VariableElement,
  {
    computedGeometry,
    computedVariables,
    elementsById,
    errors,
    disabledByGroupId,
    localVariables
  }: {
    computedGeometry: Map<ElementId, ComputedGeometry>;
    computedVariables: Map<ElementId, ComputedVariable>;
    elementsById: Map<ElementId, CadElement>;
    errors: DependencyError[];
    disabledByGroupId: Map<ElementId, ElementId>;
    localVariables: LocalVariableEvaluation;
  }
) => {
  const numeric = (value = element.expression) =>
    numericError(
      element,
      value,
      computedGeometry,
      elementsById,
      errors,
      localVariables.localVariableValues,
      localVariables.localVariableNames,
      disabledByGroupId,
      computedVariables
    );
  const point = (anchor: VariableElement["point1"], key: string) =>
    getPointAnchorOrError(
      element,
      anchor,
      key,
      computedGeometry,
      elementsById,
      errors,
      localVariables.localVariableValues,
      localVariables.localVariableNames,
      disabledByGroupId,
      computedVariables,
      Array.from(elementsById.values())
    );

  let value: number | undefined;

  if (element.valueMode === "expression") {
    value = numeric();
  }

  if (element.valueMode === "pointDistance") {
    const point1 = point(element.point1, "point1");
    const point2 = point(element.point2, "point2");
    if (point1 && point2) value = Math.hypot(point2.x - point1.x, point2.y - point1.y);
  }

  if (element.valueMode === "pointAngle") {
    const point1 = point(element.point1, "point1");
    const point2 = point(element.point2, "point2");
    if (point1 && point2) {
      value = normalizeDegrees(
        Math.atan2(point2.y - point1.y, point2.x - point1.x) * 180 / Math.PI
      );
    }
  }

  if (element.valueMode === "pointLineDistance") {
    const pointValue = point(element.point, "point");
    const line = computedLineOrError(
      element,
      element.lineId,
      computedGeometry,
      elementsById,
      errors,
      disabledByGroupId
    );
    if (pointValue && line) {
      const dx = line.end.x - line.start.x;
      const dy = line.end.y - line.start.y;
      const length = Math.hypot(dx, dy);
      if (length <= EPSILON) {
        errors.push(geometryError(element, `${line.name} は長さ0のため点線距離を計算できません。`));
      } else {
        value = Math.abs(dx * (line.start.y - pointValue.y) - (line.start.x - pointValue.x) * dy) / length;
      }
    }
  }

  if (value === undefined) return;
  computedVariables.set(element.id, {
    kind: "variable",
    elementId: element.id,
    name: element.name,
    value
  });
};

File: src-tauri/src/evaluation/variable_evaluator.rs
use serde_json::{json, Value};
use std::collections::HashMap;

use super::errors::{dependency_error, geometry_error};
use super::math::normalize_degrees;
use super::numeric_expression::evaluate_numeric_or_push;
use super::point_anchor::{point_anchor_or_error, point_from_value};
use super::types::{element_id, element_name, insert_variable, EvaluationState};

pub(crate) fn evaluate_variable_element(
    element: &Value,
    local_variables: &(HashMap<String, f64>, HashMap<String, String>),
    state: &mut EvaluationState,
) {
    let value_mode = element
        .get("valueMode")
        .and_then(Value::as_str)
        .unwrap_or("expression");
    let value = match value_mode {
        "expression" => evaluate_numeric_or_push(
            element.get("expression").unwrap_or(&Value::Null),
            state,
            element,
            &local_variables.0,
            &local_variables.1,
        ),
        "pointDistance" => {
            let Some(point1) = point_anchor_or_error(
                element,
                element.get("point1").unwrap_or(&Value::Null),
                "point1",
                state,
                &local_variables.0,
                &local_variables.1,
            ) else {
                return;
            };
            let Some(point2) = point_anchor_or_error(
                element,
                element.get("point2").unwrap_or(&Value::Null),
                "point2",
                state,
                &local_variables.0,
                &local_variables.1,
            ) else {
                return;
            };
            Some((point2.x - point1.x).hypot(point2.y - point1.y))
        }
        "pointAngle" => {
            let Some(point1) = point_anchor_or_error(
                element,
                element.get("point1").unwrap_or(&Value::Null),
                "point1",
                state,
                &local_variables.0,
                &local_variables.1,
            ) else {
                return;
            };
            let Some(point2) = point_anchor_or_error(
                element,
                element.get("point2").unwrap_or(&Value::Null),
                "point2",
                state,
                &local_variables.0,
                &local_variables.1,
            ) else {
                return;
            };
            Some(normalize_degrees(
                (point2.y - point1.y)
                    .atan2(point2.x - point1.x)
                    .to_degrees(),
            ))
        }
        "pointLineDistance" => {
            let Some(point) = point_anchor_or_error(
                element,
                element.get("point").unwrap_or(&Value::Null),
                "point",
                state,
                &local_variables.0,
                &local_variables.1,
            ) else {
                return;
            };
            let line_id = element
                .get("lineId")
                .and_then(Value::as_str)
                .unwrap_or_default();
            let Some(line) = state.computed_geometry.get(line_id).cloned() else {
                state.errors.push(dependency_error(state, element, line_id));
                return;
            };
            let Some(start) = line.get("start").and_then(point_from_value) else {
                state.errors.push(dependency_error(state, element, line_id));
                return;
            };
            let Some(end) = line.get("end").and_then(point_from_value) else {
                state.errors.push(dependency_error(state, element, line_id));
                return;
            };
            let dx = end.x - start.x;
            let dy = end.y - start.y;
            let length = dx.hypot(dy);
            if length <= 1e-9 {
                state.errors.push(geometry_error(
                    element,
                    format!(
                        "{} は長さ0のため点線距離を計算できません。",
                        element_name(element)
                    ),
                ));
                return;
            }
            Some((dx * (start.y - point.y) - (start.x - point.x) * dy).abs() / length)
        }
        _ => None,
    };

    let Some(value) = value else {
        return;
    };
    let id = element_id(element).unwrap_or_default();
    insert_variable(
        state,
        id.clone(),
        json!({
            "kind": "variable",
            "elementId": id,
            "name": element_name(element),
            "value": value
        }),
    );
}
